bench.c

```diff
@@ -0,0 +1,196 @@
+pthread_mutex_t worker_lock = PTHREAD_MUTEX_INITIALIZER;
+pthread_cond_t worker_wait = PTHREAD_COND_INITIALIZER;
+
+long time_res[MAX_THREAD] = {0};
+int idx = 0; /* for indexing `time_res` */
+FILE *bench_fd;
+
+static inline long time_diff_us(struct timeval *start, struct timeval *end)
+{
+    return ((end->tv_sec - start->tv_sec) * 1000000) +
+           (end->tv_usec - start->tv_usec);
+}
+
+static void *bomb_worker(void *arg)
+{
+    int sock_fd, rt;
+    char dummy[MAX_MSG_LEN];
+    struct timeval start, end;
+
+    /* wait until all worker created */
+    pthread_mutex_lock(&worker_lock);
+    while (!ready)
+        if ((rt = pthread_cond_wait(&worker_wait, &worker_lock))) {
+            puts("cond_wait failed!");
+            exit(-1);
+        }
+    pthread_mutex_unlock(&worker_lock);
+
+    sock_fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sock_fd == -1) {
+        goto sock_init_fail;
+    }
+
+    struct sockaddr_in info = {
+        .sin_family = PF_INET,
+        .sin_addr.s_addr = inet_addr(TARGET_HOST),
+        .sin_port = htons(TARGET_PORT),
+    };
+
+    if (connect(sock_fd, (struct sockaddr *) &info, sizeof(info)) == -1) {
+        goto connect_fail;
+    }
+
+    gettimeofday(&start, NULL);
+    send(sock_fd, msg_dum, strlen(msg_dum), 0);
+    recv(sock_fd, dummy, MAX_MSG_LEN, 0);
+    gettimeofday(&end, NULL);
+
+    close(sock_fd);
+
+    if (strncmp(msg_dum, dummy, strlen(msg_dum)))
+        goto valid_fail;
+
+    pthread_mutex_lock(&res_lock);
+    time_res[idx++] += time_diff_us(&start, &end);
+    pthread_mutex_unlock(&res_lock);
+
+    pthread_exit(NULL);
+
+sock_init_fail:
+    perror("Socket create failed");
+    exit(-1);
+
+connect_fail:
+    perror("Socket connect failed");
+    exit(-1);
+
+valid_fail:
+    puts("Validation of echo message failed...");
+    exit(-1);
+}
+
+static void create_worker(int thread_qty)
+{
+    int status;
+
+    for (int i = 0; i < thread_qty; i++) {
+        status = pthread_create(&pt[i], NULL, bomb_worker, NULL);
+        if (status) {
+            puts("Thread creation failure");
+            exit(-1);
+        }
+    }
+}
+
+static void bench(void)
+{
+    for (int i = 0; i < BENCH_COUNT; i++) {
+        ready = false;
+
+        create_worker(MAX_THREAD);
+
+        pthread_mutex_lock(&worker_lock);
+
+        ready = true;
+
+        /* all workers are ready, let's start bombing kecho */
+        pthread_cond_broadcast(&worker_wait);
+
+        pthread_mutex_unlock(&worker_lock);
+
+        /* waiting for all workers to finish the measurement */
+        for (int x = 0; x < MAX_THREAD; x++)
+            pthread_join(pt[x], NULL);
+
+        idx = 0;
+    }
+
+    for (int i = 0; i < MAX_THREAD; i++)
+        fprintf(bench_fd, "%d %ld\n", i, time_res[i] /= BENCH_COUNT);
+}
+
+int main(void)
+{
+    int status = 0;
+
+    bench_fd = fopen(RESULT_FILE_NAME, "w");
+    if (!bench_fd) {
+        status = -1;
+        goto fopen_fail;
```

Comment:
I think its confusing to use two `goto` here, why not simply print the error and return -1?
`fclose` also has undefined behavior if the parameter is an illegal pointer, which might happen after the `fopen_fail` label

Reply:
Ok, it's really confusing and dangerous, will fix it!